src/software/new_geom/vector.h

```diff
@@ -0,0 +1,511 @@
+ * @param v the Vector to subtract
+ *
+ * @return the new value of Vector u
+ */
+Vector &operator-=(Vector &u, const Vector &v);
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param s the scaling factor
+ * @param v the vector to scale
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(double s, const Vector &v) __attribute__((warn_unused_result));
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(const Vector &p, double s) __attribute__((warn_unused_result));
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator*=(Vector &p, double s);
+
+/**
+ * Divides a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scalar to divide by
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator/(const Vector &p, double s) __attribute__((warn_unused_result));
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator/=(Vector &p, double s);
+
+/**
+ * Prints a vector to a stream
+ *
+ * @param os the stream to print to
+ * @param v the Vector to print
+ *
+ * @return the stream with the Vector printed
+ */
+inline std::ostream &operator<<(std::ostream &os, const Vector &v);
+
+/**
+ * Compares two Vectors for equality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent the same vector, and false otherwise
+ */
+constexpr bool operator==(const Vector &u, const Vector &v);
+
+/**
+ * Compares two vectors for inequality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent different vectors, and false otherwise
+ */
+constexpr bool operator!=(const Vector &u, const Vector &v);
+
+inline Vector Vector::createFromAngle(const Angle &angle)
+{
+    return Vector(angle.cos(), angle.sin());
+}
+
+inline constexpr Vector::Vector() : x_(0.0), y_(0.0) {}
+
+inline constexpr Vector::Vector(double x, double y) : x_(x), y_(y) {}
+
+inline constexpr Vector::Vector(const Vector &v) : x_(v.x()), y_(v.y()) {}
+
+inline constexpr double Vector::x() const
+{
+    return x_;
+}
+
+inline constexpr double Vector::y() const
+{
+    return y_;
+}
+
+inline void Vector::set(double x, double y)
+{
+    this->x_ = x;
+    this->y_ = y;
+}
+
+inline void Vector::setX(double x)
+{
+    this->x_ = x;
+}
+
+inline void Vector::setY(double y)
+{
+    this->y_ = y;
+}
+
+inline constexpr double Vector::lensq() const
+{
+    return x_ * x_ + y_ * y_;
+}
+
+inline double Vector::len() const
+{
+    return std::hypot(x_, y_);
+}
+
+inline Vector Vector::norm() const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(x_ / len(), y_ / len());
```

Comment:
Is there a specific reason we're using `1.0e-9` here? I would imagine we'd want to compare to `EPSILON` instead (ie. check that the number is basically 0 in terms of what the computer is capable of representing).